stats/opentelemetry/internal/grpc_trace_bin_propagator_test.go

```diff
@@ -42,151 +42,195 @@ func Test(t *testing.T) {
+// its context and then verifies that same span context is extracted after
+// base64 decoding.
+func (s) TestExtract(t *testing.T) {
+	tests := []struct {
+		name string
+		sc   oteltrace.SpanContext
+		fast bool
+	}{
+		{
+			name: "fast path, valid context",
+			sc: oteltrace.SpanContext{}.WithTraceID(
+				oteltrace.TraceID{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}).WithSpanID(
+				oteltrace.SpanID{17, 18, 19, 20, 21, 22, 23, 24}).WithTraceFlags(
+				oteltrace.TraceFlags(1)).WithRemote(true),
+			fast: true,
+		},
+		{
+			name: "fast path, invalid context",
+			sc:   oteltrace.SpanContext{},
+			fast: true,
+		},
+		{
+			name: "slow path, valid context",
+			sc: oteltrace.SpanContext{}.WithTraceID(
+				oteltrace.TraceID{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}).WithSpanID(
+				oteltrace.SpanID{17, 18, 19, 20, 21, 22, 23, 24}).WithTraceFlags(
+				oteltrace.TraceFlags(1)).WithRemote(true),
+			fast: false,
+		},
+		{
+			name: "slow path, invalid context",
+			sc:   oteltrace.SpanContext{},
+			fast: false,
+		},
 	}
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	tCtx := p.Extract(ctx, c)
-	got := oteltrace.SpanContextFromContext(tCtx)
 
-	if !got.Equal(sc) {
-		t.Fatalf("got = %v, want %v", got, sc)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			p := GRPCTraceBinPropagator{}
+			bd := binary(test.sc)
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			var c otelpropagation.TextMapCarrier
+			if test.fast {
+				c = itracing.NewCustomCarrier(stats.SetIncomingTrace(ctx, bd))
+			} else {
+				c = otelpropagation.MapCarrier{itracing.GRPCTraceBinHeaderKey: base64.StdEncoding.EncodeToString(bd)}
+			}
+			tCtx := p.Extract(ctx, c)
+			got := oteltrace.SpanContextFromContext(tCtx)
+			if !got.Equal(test.sc) {
+				t.Fatalf("got = %v, want %v", got, test.sc)
+			}
+		})
 	}
 }
 
 // TestBinary verifies that the binary() function correctly serializes a valid
-// OpenTelemetry SpanContext into its binary format representation.
+// OpenTelemetry span context into its binary format representation. If span
+// context is invalid, it verifies that serialization is nil.
 func (s) TestBinary(t *testing.T) {
 	tests := []struct {
 		name string
 		sc   oteltrace.SpanContext
 		want []byte
 	}{
 		{
-			name: "valid",
+			name: "valid context",
 			sc: oteltrace.SpanContext{}.WithTraceID(
-				oteltrace.TraceID{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
-			).WithSpanID(
-				oteltrace.SpanID{17, 18, 19, 20, 21, 22, 23, 24},
-			).WithTraceFlags(
-				oteltrace.TraceFlags(1),
-			),
-			want: []byte{0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1, 17, 18, 19, 20, 21, 22, 23, 24, 2, 1},
+				oteltrace.TraceID{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}).WithSpanID(
+				oteltrace.SpanID{17, 18, 19, 20, 21, 22, 23, 24}).WithTraceFlags(
+				oteltrace.TraceFlags(1)),
```

Comment:
You could declare this as a top level variable.

Reply:
Done